File: backend/src/livekit/livekit.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { RoomServiceClient, AccessToken } from 'livekit-server-sdk';

@Injectable()
export class LivekitService {
  private readonly logger = new Logger(LivekitService.name);
  private readonly apiKey = process.env.LIVEKIT_API_KEY || 'dev-key';
  private readonly apiSecret = process.env.LIVEKIT_API_SECRET || 'dev-secret';
  private readonly livekitUrl = process.env.LIVEKIT_URL || 'wss://your-livekit-project.livekit.cloud';
  private readonly roomService: RoomServiceClient;

  constructor() {
    if (!process.env.LIVEKIT_API_KEY || !process.env.LIVEKIT_API_SECRET) {
      this.logger.warn(
        'LIVEKIT_API_KEY or LIVEKIT_API_SECRET is not set. LiveKit token generator running in dummy mode.'
      );
    }
    
    let httpUrl = this.livekitUrl.replace('wss://', 'https://').replace('ws://', 'http://');
    this.roomService = new RoomServiceClient(httpUrl, this.apiKey, this.apiSecret);
  }

  async generateAccessToken(roomName: string, participantName: string, participantIdentity?: string): Promise<string> {
    const identity = participantIdentity || `guest_${Math.floor(1000 + Math.random() * 9000)}`;

    try {
      // Create or update the room to enforce the 30 participants limit
      await this.roomService.createRoom({
        name: roomName,
        emptyTimeout: 10 * 60,
        maxParticipants: 30,
      });

      const at = new AccessToken(this.apiKey, this.apiSecret, {
        identity: identity,
        name: participantName,
        ttl: '6h',
      });
      
      at.addGrant({
        roomJoin: true,
        room: roomName,
        canPublish: true,
        canSubscribe: true,
        canPublishData: true,
      });

      const token = await at.toJwt();
      this.logger.log(`Generated LiveKit token for participant ${participantName} in room ${roomName} with max 30 participants`);
      return token;
    } catch (error: any) {
      this.logger.error(`Failed to encode LiveKit token or create room: ${error.message}`);
      throw error;
    }
  }

  getLivekitUrl(): string {
    return this.livekitUrl;
  }
}

File: backend/app/services/livekit_service.py
import time
import logging
from jose import jwt
from app.config import settings

logger = logging.getLogger("livekit_service")

class LiveKitService:
    def __init__(self):
        self.api_key = settings.LIVEKIT_API_KEY
        self.api_secret = settings.LIVEKIT_API_SECRET
        self.livekit_url = settings.LIVEKIT_URL

        if not self.api_key or not self.api_secret:
            logger.warning(
                "LIVEKIT_API_KEY or LIVEKIT_API_SECRET is not set. LiveKit token generator running in dummy mode."
            )

    def generate_access_token(self, room_name: str, participant_identity: str, participant_name: str) -> str:
        # Fallback secrets if none set in settings
        key = self.api_key or "dev-key"
        secret = self.api_secret or "dev-secret"

        now = int(time.time())
        # Token expires in 6 hours
        expire = now + 6 * 3600

        payload = {
            "iss": key,
            "sub": participant_identity,
            "name": participant_name,
            "exp": expire,
            "nbf": now,
            "video": {
                "roomJoin": True,
                "room": room_name,
                "roomList": True,
                "canPublish": True,
                "canSubscribe": True,
                "canPublishData": True
            }
        }

        try:
            token = jwt.encode(payload, secret, algorithm="HS256")
            logger.info(f"Generated LiveKit token for participant {participant_name} in room {room_name}")
            return token
        except Exception as e:
            logger.error(f"Failed to encode LiveKit token: {e}")
            raise

    def get_livekit_url(self) -> str:
        return self.livekit_url

livekit_service = LiveKitService()
